### PR TITLE
L1 prefiring weight producer 

### PR DESCRIPTION
**Pull request by lathomas**
Dear CMSSW managers/PR reviewers, 

This PR contains an EDProducer computing weights associated to the L1 ECAL prefiring issue that affected 2016 and 2017 data. 
Since the effect is not present in MC, the proposed solution is to reweight simulations by those weights. 
I add some links to documentation at the end of this post for more information.  

The code is a producer returning three double (central, up/down weights). The code loops over all photons and jets found in the event and multiplies their prefiring probabilities. 

I expect the PR to need a bit of iteration before being approved. 
In particular, I would greatly appreciate your suggestions regarding the following points: 

- Can you please provide advise regarding the code location? I created a new package but it can/should maybe enter an already existing one?
- The code relies on prefiring maps stored as TH2F in a root file. I'm not sure this is allowed... If not, can you suggest a proper format? (sorry for my ignorance). 

Please let me know if anything else is unclear,

Thanks a lot, 

Laurent  

***Documentation***
-Talk at CMS week about prefiring: 
https://indico.cern.ch/event/704625/contributions/3164334/attachments/1726987/2790093/prefiringupdateCMSweek.pdf
-Prefiring discussion at the XPOG forum: 
https://indico.cern.ch/event/764279/
-Twiki describing how to run the producer:
https://twiki.cern.ch/twiki/bin/viewauth/CMS/L1ECALPrefiringWeightRecipe
-L1 prefiring maps per object:
https://lathomas.web.cern.ch/lathomas/TSGStuff/L1Prefiring/PrefiringMaps_2016and2017/
-Validation plots: 
https://twiki.cern.ch/twiki/pub/CMS/L1ECALPrefiringWeightRecipe/prefiring_producervalidation.pdf

### CODE REVIEW

**Comment by lathomas** [comment]
@peruzzim @fgolf

**Comment by peruzzim** [comment]
If the review and backport to 10_2_X proceed in an expedite way, this could get into the next NanoAOD production, which would be very valuable.

**Review by peruzzim — COMMENTED** [review]
In order to speed up the review I give a few comments from a quick look at the code

**Comment by davidlange6** [comment]
> On Nov 30, 2018, at 2:28 PM, lathomas <notifications@github.com> wrote:
> 
> @peruzzim @davidlange6
> I tried to address all your comments and moved the code to PhysicsTools/PatUtils.
> The root file size is 15kB. Judging by the size of other files in PhysicsTools/PatUtils/data I assume this is fine (?)

it sounds like this file should just be text then  (perhaps most easily added to the python configuration, though its hard to judge how many bins are in a 15 kB root file given the overheads)?

**Comment by lathomas** [comment]
Ok then I need to check how to implement that. This will not happen before end of next week I'm afraid...
The number of bins is rather small (around 100). 
See the maps here:  https://github.com/cms-sw/cmssw/pull/25380#issuecomment-443280566 
Thank

**Comment by peruzzim** [comment]
@davidlange6 could we make an exception for this time? Or put it in a data package?
It would be valuable to have this feature integrated quickly, and the root file seems practical also for future updates.

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	 
+	 
+       }
+       //If overlapping photons have a non prefiring rate smaller than the jet, don't consider the jet in the event weight
+       //else if(nonprefiringprobfromoverlappingphotons < nonprefiringprobfromoverlappingjet ) NonPrefiringProba[fluct]*=1.;
+     
+     }
+   }
+   
+   auto NonPrefiringProb =std::make_unique <double> ( NonPrefiringProba[0]);
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: Lowercas initials

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+  dataera_ =  iConfig.getParameter<std::string>( "DataEra" ) ;
+  useEMpt_  =  iConfig.getParameter<bool>( "UseJetEMPt" ) ; 
+  prefiringRateSystUnc_ =  iConfig.getParameter<double>( "PrefiringRateSystematicUncty" ) ;
+  skipwarnings_  =  iConfig.getParameter<bool>( "SkipWarnings" ) ;
+
+  TFile *  file_prefiringmaps_;
+  std::string fname =  iConfig.getParameter<std::string>( "L1Maps" );
+  edm::FileInPath mapsfilepath("PhysicsTools/PatUtils/data/"+fname);
+  file_prefiringmaps_ = new TFile(mapsfilepath.fullPath().c_str(),"read" );
+  if(file_prefiringmaps_ == nullptr && !skipwarnings_) std::cout << "File with maps not found. All prefiring weights set to 0. " <<std::endl;
```
[the remark this answers is not in the document]
Comment (perrotta) [reply PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: (Here and everywhere below....)

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+       
+       double ptem_jet =  pt_jet*(jet.neutralEmEnergyFraction()+jet.chargedEmEnergyFraction());
+       double prefiringprob_jet = (useEMpt_) ? GetPrefiringRate( eta_jet, ptem_jet, h_prefmap_jet , fluct) : GetPrefiringRate( eta_jet, pt_jet , h_prefmap_jet , fluct);
+       //useEMpt =true if one wants to use maps parametrized vs Jet EM pt instead of pt.
+       double nonprefiringprobfromoverlappingjet =(1.-prefiringprob_jet);
+       //If there are no overlapping photons, just multiply by the jet non prefiring rate
+       if(nonprefiringprobfromoverlappingphotons ==1.)    NonPrefiringProba[fluct]*= (1.-prefiringprob_jet);
+       //If overlapping photons have a non prefiring rate larger than the jet, then replace these weights by the jet one
+       else if(nonprefiringprobfromoverlappingphotons > nonprefiringprobfromoverlappingjet ) {
+	 if(nonprefiringprobfromoverlappingphotons !=0.)NonPrefiringProba[fluct]*= nonprefiringprobfromoverlappingjet /nonprefiringprobfromoverlappingphotons;
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: indent

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+  if(h_prefmap==nullptr ) return 0.;
+  //Check pt is not above map overflow
+  int nbinsy = h_prefmap->GetNbinsY();
+  double maxy= h_prefmap->GetYaxis()->GetBinLowEdge(nbinsy+1);
+  if(pt>=maxy) pt = maxy-0.01;
+  int thebin= h_prefmap->FindBin(eta,pt);
+  
+  double prefrate =  h_prefmap->GetBinContent(thebin);
+  
+  if(fluctuation == up) prefrate = TMath::Min(TMath::Max(prefrate +  h_prefmap->GetBinError(thebin), (1.+prefiringRateSystUnc_)*prefrate),1.);
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: Use  ```std::min```,  ```std::max```

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+  if(fluctuation == down) prefrate = TMath::Max(TMath::Min(prefrate -  h_prefmap->GetBinError(thebin), (1.-prefiringRateSystUnc_)*prefrate),0.);    
+  
+  return prefrate;
+
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: remove one empty line

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+	 double prefiringprob_gam =  GetPrefiringRate( eta_gam, pt_gam, h_prefmap_photon , fluct);
+	 nonprefiringprobfromoverlappingphotons  *= (1.-prefiringprob_gam) ;
+       }
+       
+       double ptem_jet =  pt_jet*(jet.neutralEmEnergyFraction()+jet.chargedEmEnergyFraction());
+       double prefiringprob_jet = (useEMpt_) ? GetPrefiringRate( eta_jet, ptem_jet, h_prefmap_jet , fluct) : GetPrefiringRate( eta_jet, pt_jet , h_prefmap_jet , fluct);
+       //useEMpt =true if one wants to use maps parametrized vs Jet EM pt instead of pt.
+       double nonprefiringprobfromoverlappingjet =(1.-prefiringprob_jet);
+       //If there are no overlapping photons, just multiply by the jet non prefiring rate
+       if(nonprefiringprobfromoverlappingphotons ==1.)    NonPrefiringProba[fluct]*= (1.-prefiringprob_jet);
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]:
```
if(nonprefiringprobfromoverlappingphotons ==1.)  NonPrefiringProba[fluct] *= nonprefiringprobfromoverlappingjet;
```

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc (hunk 3 above)
Comment (perrotta) [reply PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: ... and add a few spaces to improve reading, e.g.
```
	 if (nonprefiringprobfromoverlappingphotons !=0.) 
             NonPrefiringProba[fluct] *= nonprefiringprobfromoverlappingjet / nonprefiringprobfromoverlappingphotons;
```

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+  void endStream() override;
+  virtual double GetPrefiringRate( double eta, double pt, TH2F * h_prefmap, fluctuations fluctuation);
+  
+  edm::InputTag srcPhotons_;
+  edm::EDGetTokenT<std::vector< pat::Photon> >  photons_token_; 
+  
+  edm::InputTag srcJets_;
+  edm::EDGetTokenT<std::vector< pat::Jet> > jets_token_;
+
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: remove one empty line

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  
+private:
+  void beginStream(edm::StreamID) override;
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  void endStream() override;
+  virtual double GetPrefiringRate( double eta, double pt, TH2F * h_prefmap, fluctuations fluctuation);
+  
+  edm::InputTag srcPhotons_;
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: Where is this  ```srcPhotons_``` needed?
Please remove

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+  edm::EDGetTokenT<std::vector< pat::Photon> >  photons_token_; 
+  
+  edm::InputTag srcJets_;
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: Where is this  ```srcJets_``` needed?
Please remove

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+  edm::FileInPath mapsfilepath("PhysicsTools/PatUtils/data/"+fname);
+  file_prefiringmaps_ = new TFile(mapsfilepath.fullPath().c_str(),"read" );
+  if(file_prefiringmaps_ == nullptr && !skipwarnings_) std::cout << "File with maps not found. All prefiring weights set to 0. " <<std::endl;
+  TString mapphotonfullname= "L1prefiring_photonptvseta_"+ dataera_; 
+  if(!file_prefiringmaps_->Get(mapphotonfullname)&& !skipwarnings_)std::cout << "Photon map not found. All photons prefiring weights set to 0. " <<std::endl; 
+  h_prefmap_photon =(TH2F*) file_prefiringmaps_->Get(mapphotonfullname);
+  TString mapjetfullname= (useEMpt_) ? "L1prefiring_jetemptvseta_"+ dataera_  :  "L1prefiring_jetptvseta_"+ dataera_ ; 
+  if(!file_prefiringmaps_->Get(mapjetfullname)&& !skipwarnings_) std::cout << "Jet map not found. All jets prefiring weights set to 0. "<< std::endl ; 
+  h_prefmap_jet =(TH2F*) file_prefiringmaps_->Get(mapjetfullname);
+  file_prefiringmaps_->Close();
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: Does ```Close()``` delete it?
Otherwise, please explicitely delete ```file_prefiringmaps_```

**Comment by perrotta** [comment]
please test

**Comment by lathomas** [comment]
Hi. 
So the PR has been updated to address the various comments. 
Please let me know if anything else is needed on my side ! 
Thanks !

**Comment by perrotta** [comment]
Thank you @lathomas for the updates.

I think there is still to solve the issue of the root file which now needs to be stored.

And moreover, I would like to understand how it is supposed to enter the release. Up to now, the producer is provided unpluged from the rest of the code: don't you plan to have it included in the future miniAOD (e.g. ultra-legacy) and nanoAOD productions, for example?

**Comment by lathomas** [comment]
Thanks for the feedback, 

Regarding the root file, we were wondering if for the short term and in order to speed up things, the file could remain as is since it's not too big (15kB). 
I can work on converting this to text file but that will require some time. 
In any case, I'll work on converting this to text file in the near future. 

About plugging the module to the sequence, I think @peruzzim  wanted to do it in a separate PR. 
Basically the hope was to have this integrated fast so that it could be added to the next NANOAOD production.

**Comment by lathomas** [comment]
@perrotta 
We would really like to avoid converting the root file to a text file as this will make the code and possible updates more complicated. Is there maybe a database where this file can be stored (independently from git)? 

Thanks !

**Comment by perrotta** [comment]
@lathomas the cleanest way is to submit PhysicsTools/PatUtils/data/L1PrefiringMaps.root as an external (and it can be a root file, then) and remove the corresponding file from this pull request

**Comment by arizzi** [comment]
@perrotta isn't 15k  small enough to be in the noise and  not deserving a new cms-data package? My understanding is that the problem is not really the format (root vs text) in general but rather just the size.

@lathomas there is a database but I'm afraid you would need a new condformat taking even more time to finally integrate.

btw are you sure is it one 100 bins histo producing a 15kb root files? root overhead should be less than that

**Comment by lathomas** [comment]
@arizzi 
The file currently contains 6 histograms (TH2F) with 416 bins in total. Most of the bins are actually empty and can be merged. That can help reduce the size. I'll give that a try tomorrow.

**Comment by lathomas** [comment]
So I did try to merge all empty bins together and the size reduction is really 0. 
For the record, there's already a 4kB root file in PhysicsTools/PatUtils/data (pfJetResolutionMCtoDataCorrLUT.root) that contains one TH2D with 5 bins.

**Comment by perrotta** [comment]
@lathomas @arizzi 
My impression is that it can be much faster to open a PR for the external than having to discuss it. However, if you believe it is important keeping your root file inside CMSSW, and since I personally don't care so much about where it gets stored, let ask the release gurus.
@davidlange6 @fabiocos : do you think that the 15 kB root file in this PR can remain inside CMSSW, or should it be more conveniently moved to an external (perhaps together with the other 4 kB root file that @lathomas discovered in PhysicsTools/PatUtils/data )? Please let us know.

**Comment by arizzi** [comment]
@lathomas I looked at the file and I read that histo names have the run era (2016) inside. Is it in the plan to add more of those for 2017, 2018 etc? If so, i.e. if the file may grow, probably it is best to move to the external. 
Anyhow my 2cents if we have to rush to produce a release to run before xmas is to leave this for now in and fix on master in the longer term (but I let O&C people comment on this)

**Comment by davidlange6** [comment]
personally I don't think this should even be a root file as I commented before... but yes, binary files should be in externals.

**Comment by lathomas** [comment]
@arizzi This code emulates the L1 prefiring that only affected 2016/2017 data so there will not be additional maps for 2018. There's no plan to update the file in the short term with additional maps. 
Can someone provide guidance on how to make this root file external? 

Thanks !

**Comment by fabiocos** [comment]
@lathomas 15 kB are not the 50 MB that gave us recently problems in the history, but since we are discussing how to limit the git repository size I think we should clean these things away since the beginning. @smuzaffar may help you in getting the external data package, especially if more has to come, as @arizzi was mentioning

**Comment by lathomas** [comment]
It looks that we are not far from converging so it would be great to get this integrated before the MC NANOAOD production starts (right after xmas break from what I understand).

The remaining item is basically moving the root file to an external package. 
@fabiocos @smuzaffar 
Indeed some help/instructions to do that would be welcome as I'm not familiar with this.

Thanks a lot !

**Comment by arizzi** [comment]
@perrotta isn't  it transparent for the code?

**Comment by perrotta** [comment]
@arizzi : it should, but I would check it explicitly

**Comment by perrotta** [comment]
@fabiocos @smuzaffar @mrodozov could you please either provide a repository in the externals for
```
PhysicsTools/PatUtils/data/L1PrefiringMaps.root
PhysicsTools/PatUtils/data/pfJetResolutionMCtoDataCorrLUT.root
```
or let us know which is the correct procedure that has to be followed for that?

Once that can get fixed, I see no other showstopper for approving and eventually merge this PR

**Comment by smuzaffar** [comment]
@fabiocos @perrotta , I have created https://github.com/cms-data/PhysicsTools-PatUtils . Before I upload the data files, I see PhysicsTools/PatUtils/data also contains some txt files. Should we move all of these data files in to external? Although deleting thexe txt files is not going to reduce the git repo size but it will reduce the cmssw distribution size.

**Comment by lathomas** [comment]
@smuzaffar @fabiocos @perrotta 
Thanks a lot for helping on this. 
Once the file is moved to external, I assume the following line [1] will need to be updated. 
I'm not sure how though. 
Thanks ! 

[1] https://github.com/cms-sw/cmssw/blob/5d7ed9e0ac13070d7e9469490f7ee51f7bf6af04/PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc#L75

**Comment by smuzaffar** [comment]
@lathomas , no you do not need to update that. The file relative path will remain the same.

**Comment by lathomas** [comment]
Ah even better. Thanks and sorry for the noise.

**Comment by perrotta** [comment]
Thank you @smuzaffar ! About moving also the txt files I have no strong opinion: if you believe it can be worth doing so, please do
@lathomas : you don't have to update any line of code, but please remove from the PR all the files which are going to be moved in the externals

**Comment by lathomas** [comment]
@perrotta The root files have been removed as requested. I can remove the text files as well in a separate PR if needed but I prefer to get this one finalized as soon as possible. Thanks !

**Comment by smuzaffar** [comment]
cmsdist PR https://github.com/cms-sw/cmsdist/pull/4595 contains all the data files of PhysicsTools/PatUtils/data.
@lathomas, as mentioned here https://github.com/cms-sw/cmssw/pull/25380#issuecomment-448027088 you need to cleanup the history of this PR to not contain any reference to these "added", "deleted" files.

**Comment by perrotta** [comment]
please test with  cms-sw/cmsdist#4595

**Comment by perrotta** [comment]
+1
- Analysis tool not used by default in any official workflow, but meant for now to private analyses and eventually possibly also nanoAOD production
- By following the recipe provided, it can run (e.g. on nanoAOD)
- The code reflects what declared, and most of what sorted out in the discussion in the thread
- It could be easily transformed into a global module, instead of being a stream one: but this can be left to a next iteration, when also all data files still in ```PhysicsTools/PatUtils/data``` should be removed from the CMSSW repository (since they are now all also copied into the external directory) 
- Jenkins tests pass and show no differences

**Comment by lathomas** [comment]
Hi, 
Thanks @perrotta for the sign off. It would be great to finalize this before the MC NANOAOD production starts.
Can I proceed with the backport (for 10_2_X, 9_4_X and possibly also 8_0_X?)? 
Thanks !

File: PhysicsTools/PatUtils/python/L1ECALPrefiringWeightProducer_cfi.py [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import FWCore.ParameterSet.Config as cms
+
+prefiringweight = cms.EDProducer("L1ECALPrefiringWeightProducer",
```
Comment (fabiocos) [inline PhysicsTools/PatUtils/python/L1ECALPrefiringWeightProducer_cfi.py]: @lathomas @perrotta this line seems inconsistent with

descriptions.add("l1ECALPrefiringWeightProducer",desc);

as it should be 

descriptions.add("prefiringweight", desc); 

what is the reason for this discrepancy? And in general to have both the fillDescriptions and the external cfi?

**Comment by perrotta** [comment]
@lathomas : please modify L1ECALPrefiringWeightProducer_cfi.py (and rename it "_cff") so that it imports from l1ECALPrefiringWeightProducer_cfi.py and only modifies the two different parameter values that you are changing here, as pointed out by @fabiocos in https://github.com/cms-sw/cmssw/pull/25380#discussion_r245685665

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 12]
```diff
@@ -214,12 +214,12 @@ L1ECALPrefiringWeightProducer::fillDescriptions(edm::ConfigurationDescriptions&
   desc.add<edm::InputTag>("TheJets", edm::InputTag("slimmedJets"));
   desc.add<std::string>("L1Maps","L1PrefiringMaps.root");
   desc.add<std::string>("DataEra", "2017BtoF");
-  desc.add<bool>("UseJetEMPt", true);
+  desc.add<bool>("UseJetEMPt", false);
   desc.add<double>("PrefiringRateSystematicUncty",0.2);
-  desc.add<bool>("SkipWarnings", true);
-  
-  descriptions.add("l1ECALPrefiringWeightProducer",desc);
+  desc.add<bool>("SkipWarnings", false);
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: Why did you change the default value into this more verbose one? I think it can be safer to have the "SkipWarnings" true by default, and only enable warnings if needed

File: PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc [hunk 13]
```diff
@@ -214,12 +214,12 @@ L1ECALPrefiringWeightProducer::fillDescriptions(edm::ConfigurationDescriptions&
...
 
+  descriptions.add("prefiringweight", desc);
```
Comment (perrotta) [inline PhysicsTools/PatUtils/plugins/L1ECALPrefiringWeightProducer.cc]: It is safer to have a default config with a generic name, as you had before ("l1ECALPrefiringWeightProducer") and clone and modify the specialization ("prefiringweight") in the cff file (PhysicsTools/PatUtils/python/L1ECALPrefiringWeightProducer_cff.py), or whatever other config users want to use for it.

File: PhysicsTools/PatUtils/python/L1ECALPrefiringWeightProducer_cff.py [hunk 14]
```diff
@@ -0,0 +1,4 @@
+import FWCore.ParameterSet.Config as cms
+
+process.load('PhysicsTools.PatUtils.l1ECALPrefiringWeightProducer_cfi')
```
Comment (perrotta) [inline PhysicsTools/PatUtils/python/L1ECALPrefiringWeightProducer_cff.py]: Clone and modify (if needed, otherwise just clone, see above)

**Comment by perrotta** [comment]
@lathomas : before launching the tests, could you please either answer or cope with the two still unanswered comments in https://github.com/cms-sw/cmssw/pull/25380/files#r238603356 and https://github.com/cms-sw/cmssw/pull/25380/files#r238632854
(Sorry for having overlooked them before)

**Comment by lathomas** [comment]
@perrotta done !

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
@lathomas : do you have an analyzer (which runs on nanoAOD, possibly) that consumes these weights? I.e. something obtained with https://twiki.cern.ch/twiki/bin/viewauth/CMS/L1ECALPrefiringWeightRecipe but avoiding me the need to recreate everything from scratch following that recipe.
Once I can verify that this PR can run correctly even after the last updates I will sign it.

**Comment by perrotta** [comment]
+1
- Already signed (see https://github.com/cms-sw/cmssw/pull/25380#issuecomment-448249994), then cfg fixed, rebased, and a few further adjustments were applied following the code review
- This is an analysis tool not used by default in any official workflow, but meant for now to private analyses and eventually possibly also nanoAOD production
- By following the recipe provided, it runs (I managed to do it on nanoAOD)
- The code reflects what declared, and subsequent review comments
- Jenkins tests pass and show no differences
- Needed external got already merged in the master

**Comment by lathomas** [comment]
@perrotta : thanks !

**Comment by fabiocos** [comment]
+1

the configuration of the module has been properly adjusted, thanks

**Comment by fabiocos** [comment]
merge

**Comment by lathomas** [comment]
@fabiocos  @perrotta @peruzzim 
Thanks for the merging and for your help.
Can I proceed with the backport? What branches should I consider? I guess at least 94X is needed but are there others?

**Comment by perrotta** [comment]
Analyses will be performed in 10_2_X (preferably), and also in 94X:  
please consider backporting in both cycles.

lathomas <notifications@github.com> ha scritto:

**Comment by mrodozov** [comment]
@perrotta @lathomas 
https://github.com/cms-sw/cmsdist/pull/4635
https://github.com/cms-sw/cmsdist/pull/4636
are for 10_2 and 9_4

**Comment by perrotta** [comment]
Thank you Mircho!

Mircho Rodozov <notifications@github.com> ha scritto:
